src/lib/Algebra.hs

```diff
@@ -243,16 +243,21 @@ blockAsCPoly (Block _ decls' result') =
     indexAsCPoly = \case
       Var v                       -> varAsCPoly v
       Con (FinVal _ i)            -> intAsCPoly i
-      Con (IndexRangeVal _ _ _ i) -> intAsCPoly i
       _                           -> empty
 
     intAsCPoly :: (SubstReader CPolySubstVal m, Alternative2 m)
                 => Atom i -> m i o (ClampPolynomial o)
     intAsCPoly = \case
       Var v       -> varAsCPoly v
+      -- XXX: this case is needed to handle the Nat-wrapper types, RangeFrom,
+      -- RangeFromExc, RangeTo, and RangeToExc. But user-defined index sets
+      -- won't always just be wrappers around `Nat`! So we need Algebra to be
+      -- aware of their definitions.
+      ProjectElt (0:|[]) v -> varAsCPoly v
```

Comment:
I'm pretty sure that this will produce ill-typed IR. We might need to extend the set of valid polynomial vars to additionally include projections applied to them. That is, rather than normalizing the `ProjectElt` case to the `Var` case, normalize `Var` to an empty projection. Then, once you evaluate the polynomial, reconstruct the appropriate projections.

Reply:
Yeah, it's totally broken. As we discussed off-thread, I'll try to fix it by making Algebra more Ix-aware in general. I'll do that in a PR on top of this one and then we can merge both together.